From 6b71da95582d1c64aad6948ce37beeb39e05ed9e Mon Sep 17 00:00:00 2001
From: Michael Kruse <isl@meinersbur.de>
Date: Mon, 16 Nov 2020 20:57:16 +0000
Subject: [PATCH 1/2] isl_mat_alloc: fix undefined null pointer arithmetic

Arithmetic on null pointers is undefined behaviour, even if the
resulting pointer is never dereferenced. If isl_mat_alloc is called with
zero n_col argument, but nonzero n_row, the allocated isl_blk size is
still zero and isl_blk_alloc returns an empty allocation with the data
member set to NULL. Because n_row is nonzero, it is still used to
initialize the isl_mat row addresses using an offset to NULL isl_blk
data member.

This is fixed by initializing the row array with NULL using calloc and
only conditionally assigning the pointer elements to an offset derived
from the isl_blk allocation.

Signed-off-by: Michael Kruse <isl@meinersbur.de>
---
Aithmetic on the NULL-pointer is undefined behaviour in C (C17 6.5.6p8)
as well as C++ (http://eel.is/c++draft/expr.add#4).

The occurance is diagnosed by UndefinedBehaviourSanitizer, when running
e.g. isl_test, shown below. The same problem probably also applies to
isl_mat_extend, but does not occur during a normal test run. Another
possibility to fix the problem would be to make isl_blk_alloc return a
malloc(1) allocation, with the same rationale autoconf's AC_FUNC_MALLOC
ensures that malloc(0) returns non-NULL

$ UBSAN_OPTIONS=report_error_type=1,print_stacktrace=1 ./isl_test
[...]
conversion
/home/meinersbur/src/isl/isl_mat.c:72:33: runtime error: applying zero offset to null pointer
    #0 0x7f00d6e13061 in isl_mat_alloc /home/meinersbur/src/isl/isl_mat.c:72:33
    #1 0x7f00d6d753c5 in particular_solution /home/meinersbur/src/isl/isl_equalities.c:105:9
    #2 0x7f00d6d753c5 in isl_mat_parameter_compression /home/meinersbur/src/isl/isl_equalities.c:320:8
    #3 0x7f00d6df98ae in normalize_divs /home/meinersbur/src/isl/isl_map_simplify.c:1027:6
    #4 0x7f00d6df98ae in isl_basic_map_simplify /home/meinersbur/src/isl/isl_map_simplify.c:1539:10
    #5 0x7f00d6e0fc00 in basic_map_collect_diff /home/meinersbur/src/isl/isl_map_subtract.c:369:9
    #6 0x7f00d6e0f04c in basic_map_diff_is_empty /home/meinersbur/src/isl/isl_map_subtract.c:683:6
    #7 0x7f00d6e0f04c in map_diff_is_empty /home/meinersbur/src/isl/isl_map_subtract.c:704:14
    #8 0x7f00d6e0f04c in map_is_subset /home/meinersbur/src/isl/isl_map_subtract.c:871:14
    #9 0x7f00d6dd6880 in map_is_equal /home/meinersbur/src/isl/isl_map.c:9091:14
    #10 0x446605 in test_set_conversion /home/meinersbur/src/isl/isl_test.c:9093:11
    #11 0x446605 in test_conversion /home/meinersbur/src/isl/isl_test.c:9240:6
    #12 0x450bcf in main /home/meinersbur/src/isl/isl_test.c:10930:7
    #13 0x7f00d5b9d349 in __libc_start_main (/lib64/libc.so.6+0x24349)
    #14 0x411369 in _start /home/abuild/rpmbuild/BUILD/glibc-2.26/csu/../sysdeps/x86_64/start.S:120

SUMMARY: UndefinedBehaviorSanitizer: nullptr-with-offset /home/meinersbur/src/isl/isl_mat.c:72:33 in

$ ./isl_test --version
isl-0.23-2-g4e72b588-IMath-32


 isl_mat.c | 8 +++++---
 1 file changed, 5 insertions(+), 3 deletions(-)

diff --git a/isl_mat.c b/isl_mat.c
index 93622b97..f5500045 100644
--- a/isl_mat.c
+++ b/isl_mat.c
@@ -64,12 +64,14 @@ __isl_give isl_mat *isl_mat_alloc(isl_ctx *ctx,
 	mat->block = isl_blk_alloc(ctx, n_row * n_col);
 	if (isl_blk_is_error(mat->block))
 		goto error;
-	mat->row = isl_alloc_array(ctx, isl_int *, n_row);
+	mat->row = isl_calloc_array(ctx, isl_int *, n_row);
 	if (n_row && !mat->row)
 		goto error;
 
-	for (i = 0; i < n_row; ++i)
-		mat->row[i] = mat->block.data + i * n_col;
+	if (mat->block.data) {
+		for (i = 0; i < n_row; ++i)
+			mat->row[i] = mat->block.data + i * n_col;
+	}
 
 	mat->ctx = ctx;
 	isl_ctx_ref(ctx);
-- 
2.27.0

